### PR TITLE
[Siew Hui Zhuan] iP

### PR DESCRIPTION
**Pull request by huizhuansam**
# Pull Request 
This is my pull request. Look at it! So many commits! It must be an *amazing* pull request then. 

## What's Good 👍🏼 
- Search tasks by keyword
- Search tasks by date
- Save tasks to a file so you can quit the program, start it up and **boom** your tasks are still there (\**click*\* **nice**)

## Things to Note 📓 
1. This program might be buggy
2. I absolutely loathed writing the JUnit tests
3. **2.** will be the reason for **1.** 

## Impressive Stuff 🥵 
`TaskList.java` line 118 - 132, where I attempted to write a functional flavored class method:
```java
/**
 * Returns a TaskList containing Task that satisfy the supplied predicate.
 *
 * @param predicate The predicate to test the Task with.
 * @return A TaskList containing Task that satisfies the given predicate.
 */
public TaskList filter(Function<Task, Boolean> predicate) {
    ArrayList<Task> res = new ArrayList<>();
    for (int i = 0; i < size(); i++) {
        if (predicate.apply(list.get(i))) {
            res.add(list.get(i));
        }
    }
    return new TaskList(res);
}
```
so that I can write the find task method in `FindCommand.java` as such:
```java
/**
 * Prints the tasks matching the keyword.
 *
 * @param tasks The list of tasks.
 * @param ui The UI object.
 * @param storage The storage utility.
 */
@Override
public void execute(TaskList tasks, Ui ui, Storage storage) {
    ui.print("Here are the tasks containing \"" + keyword + "\":");
    ui.print(tasks.filter((task) -> task.keywordMatcher(keyword)).toStringArray());
}
```
Functional programming is so fun! ~~Java isn't, though~~.

## Todo ✅ 
- [x] Write this PR description
- [x] Finish week 4 IP

## Apology Video (Gone Wrong) (Gone Sussy) 🙇🏼  😢  
[Link to my apology video](https://www.youtube.com/watch?v=dQw4w9WgXcQ)

### CODE REVIEW

File: src/main/java/duke/command/Command.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package duke.command;
+
+import java.io.IOException;
+
+import duke.exception.NoSuchTaskException;
+import duke.util.Storage;
+import duke.util.TaskList;
+
+/** An abstract class representing the user's commands. */
+public abstract class Command {
```
Comment (bharathcs) [inline src/main/java/duke/command/Command.java:10]: I like the way you've structured the Command class, the pure functional form is great.

File: src/main/java/duke/command/DeleteCommand.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+import duke.exception.BadInputFormatException;
+import duke.exception.NoSuchTaskException;
+import duke.util.Storage;
+import duke.util.TaskList;
+
+/** Represents the "delete" command. */
+public class DeleteCommand extends Command {
+    /** The index of the Task to be deleted. */
+    private int index;
```
Comment (bharathcs) [inline src/main/java/duke/command/DeleteCommand.java]: Just to make things clearer (for devs), maybe this should be final? Not that it really matters, pretty sure the compiler will optimise for it since its effectively final.

Comment (bharathcs) [reply src/main/java/duke/command/DeleteCommand.java]: I've also noticed this is other Command files (indices and dates), so do remember to check those out if you want to move on this suggestion.

File: src/main/java/duke/command/Command.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+     *
+     * @param tasks The list of tasks in the program.
+     * @param storage The storage utility.
+     * @throws IOException If there is an IO exception.
+     * @throws NoSuchTaskException If no tasks are found.
+     */
+    public abstract String execute(TaskList tasks, Storage storage) throws IOException, NoSuchTaskException;
+
+    /** Returns true if the Command is an exit command, else false.*/
+    public abstract boolean isExit();
```
Comment (bharathcs) [inline src/main/java/duke/command/Command.java:23]: Makes sense, I like this!

File: src/main/java/duke/task/Event.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
+package duke.task;
+
+import org.json.simple.JSONObject;
+
+import duke.date.Date;
+import duke.exception.BadInputFormatException;
+import duke.exception.InvalidDateException;
+
+/** Represents an Event object. */
+public class Event extends DatedTask {
```
Comment (bharathcs) [inline src/main/java/duke/task/Event.java:10]: Hey, I think Events are supposed to have two dates (a to and from), and thats what differentiates them from deadline.

See [level 4 (Todos, Events, Deadlines)](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/projectDuke/index.html#level-4-todos-events-deadlines)

File: src/main/java/duke/task/Task.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+     * @param description the description of the task
+     * @param isDone the status of the task
+     */
+    protected Task(String description, boolean isDone) {
+        this.description = description;
+        this.isDone = isDone;
+    }
+
+    /** Marks the task's status as done */
+    public Task markTaskAsDone() {
```
Comment (bharathcs) [inline src/main/java/duke/task/Task.java]: Just thinking, would it be neater if it was just `taskObj.setDone()` instead of `taskObj.markTaskAsDone()`?

File: src/main/java/duke/task/Task.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    /**
+     * Matches a keyword with the description of the Task. The keyword must match a whole word in the description.
+     *
+     * @param keyword The keyword as a string.
+     * @return True if the description matches the keyword, false otherwise.
+     */
+    public final boolean matchKeyword(String keyword) {
+        return Pattern.compile(".*\\b" + keyword + "\\b.*").matcher(description).find();
```
Comment (bharathcs) [inline src/main/java/duke/task/Task.java:94]: Nicely done

File: src/main/java/duke/util/Parser.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+import duke.command.ExitCommand;
+import duke.command.FindCommand;
+import duke.command.ListCommand;
+import duke.enums.Commands;
+import duke.exception.DukeException;
+import duke.exception.EmptyCommandException;
+import duke.exception.UnknownCommandException;
+
+/** Utility class that handles parsing of user input to program command. */
+public class Parser {
```
Comment (bharathcs) [inline src/main/java/duke/util/Parser.java:17]: I really like how you did this, neat implementation.

File: src/main/java/duke/util/Storage.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+            throw new FileNotFoundException();
+        }
+    }
+
+    /**
+     * Writes the tasks to a JSON file in the specified file path.
+     *
+     * @param tasks The TaskList created from the instance of the Duke program.
+     */
+    @SuppressWarnings("unchecked")
```
Comment (bharathcs) [inline src/main/java/duke/util/Storage.java]: Perhaps you should clarify in your javadocs on the reason for your warning suppression? (Note: I've seen this in a few other places too)

File: src/main/java/duke/util/TaskList.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    /**
+     * Returns an array of string representation of the Tasks in the list, numbered from 1.
+     *
+     * @return An array of string representation of the Tasks in the list, numbered from 1.
+     */
+    public String[] toStringArray() {
+        String[] copy = list.stream().map(Object::toString).toArray(String[]::new);
+        for (int i = 0; i < size(); i++) {
+            copy[i] = (i + 1) + "." + copy[i];
+        }
```
Comment (bharathcs) [inline src/main/java/duke/util/TaskList.java]: Nice! Unfortunate we don't have combine for Java streams or you could use stream operations for the indexing too.

**Review by bharathcs — COMMENTED** [review]
LGTM! You've done a great job, and I love the functional style of your code.

> "Nice and spicy"
> -- Siew Hui Zhuan

File: src/main/java/duke/util/Storage.java (hunk 8 above)
Comment (huizhuansam) [reply src/main/java/duke/util/Storage.java]: Good call. The warning suppression is because I can guarantee the correctness of the object type during runtime (line 73 of this file), but maybe other devs might get confused. I'll leave an explanatory comment here.

File: src/main/java/duke/task/Task.java (hunk 5 above)
Comment (huizhuansam) [reply src/main/java/duke/task/Task.java]: In my opinion, `markTaskAsDone()` sounds more natural than `setDone()`, as it is more akin to real-life task lists that you will "mark" tasks as done. But I understand `setDone()` sounds more OO because getters and setters or whatnot. Will consider changing this for the sake of other devs looking at my code. 👍🏼

File: src/main/java/duke/command/DeleteCommand.java (hunk 2 above)
Comment (huizhuansam) [reply src/main/java/duke/command/DeleteCommand.java]: You are right about this, but I think it is not terribly important to fix this now. Probably will do codebase optimization towards the end of the project, I'll take this into account. Thanks for the heads up.

File: src/main/java/duke/task/Event.java (hunk 4 above)
Comment (huizhuansam) [reply src/main/java/duke/task/Event.java:10]: Nice catch, I somehow missed this part.

File: src/main/java/duke/util/TaskList.java (hunk 9 above)
Comment (huizhuansam) [reply src/main/java/duke/util/TaskList.java]: My feelings exactly... the code would've looked neater without the for loop, but it is Java, and with it comes clumsy syntax :(

**Comment by huizhuansam** [comment]
@bharathcs Thanks for the review! You really pored through my code and picked out the sketchiest bits haha. Good eye for detail!

File: src/main/java/duke/command/DoneCommand.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+            throw new BadInputFormatException();
+        }
+    }
+
+    /**
+     * Marks the Task at the specified index as done.
+     *
+     * @param tasks The list of tasks in the program.
+     * @param storage The storage utility.
+     * @throws NoSuchTaskException If the task does not exist at the specified index.
```
Comment (Neha-5678) [inline src/main/java/duke/command/DoneCommand.java:43]: Maybe you could consider adding a description of what the method returns under Javadocs? I think you may have missed it out! You could consider making this change to all your command classes (classes that inherit from Command class) :)

File: src/main/java/duke/command/ListCommand.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
+package duke.command;
+
+import duke.date.Date;
```
Comment (Neha-5678) [inline src/main/java/duke/command/ListCommand.java:3]: Maybe you could consider standardising the order of the import statements across your files to make it easier to browse through? Your code is really clean so I am picking on the minute details.

File: src/main/java/duke/enums/Tasks.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
+package duke.enums;
```
Comment (Neha-5678) [inline src/main/java/duke/enums/Tasks.java:1]: You already did an amazing job for Javadocs and I am aware that the requirement was to add documentation to half of the files, however I just felt that it would be even better to add some documentation to this file if time persists so that it is easier to understand the code.

File: src/main/java/duke/exception/FileNotFoundException.java [hunk 13]
```diff
@@ -0,0 +1,4 @@
+package duke.exception;
```
Comment (Neha-5678) [inline src/main/java/duke/exception/FileNotFoundException.java:1]: Is this file necessary? Would just like to check!

File: src/main/java/duke/task/Deadline.java [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    public static Deadline of(String description) throws BadInputFormatException, InvalidDateException {
+        return new Deadline(description);
+    }
+
+    /**
+     * Factory Deadline method.
+     *
+     * @param description The Deadline's description.
+     * @param by The date to complete the Task by.
+     * @param isDone The done status of the Task
```
Comment (Neha-5678) [inline src/main/java/duke/task/Deadline.java]: I think you may have forgotten to add a punctuation behind this parameter description. Maybe you could add it in.

File: src/main/java/duke/task/Event.java [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+    /**
+     * Event constructor.
+     *
+     * @param description The event's description
```
Comment (Neha-5678) [inline src/main/java/duke/task/Event.java]: Just a reminder to add punctuation for these parameter descriptions :)

**Review by Neha-5678 — COMMENTED** [review]
You code is very neat and it was a very difficult task to spot coding violations. Good job! just a few nits to fix :-)

File: src/main/java/duke/command/DoneCommand.java (hunk 10 above)
Comment (huizhuansam) [reply src/main/java/duke/command/DoneCommand.java:43]: Ah, I accidentally left this out after implementing my GUI. Thanks for pointing it out :)

File: src/main/java/duke/command/ListCommand.java (hunk 11 above)
Comment (huizhuansam) [reply src/main/java/duke/command/ListCommand.java:3]: Can I clarify what standard are you referring to? IntelliJ and CheckStyle seem to disagree, so I went with CheckStyle's import rules (if you are wondering).

File: src/main/java/duke/enums/Tasks.java (hunk 12 above)
Comment (huizhuansam) [reply src/main/java/duke/enums/Tasks.java:1]: Noted, I will add more Javadocs if I have the time.

File: src/main/java/duke/exception/FileNotFoundException.java (hunk 13 above)
Comment (huizhuansam) [reply src/main/java/duke/exception/FileNotFoundException.java:1]: Yep! This exception is used in `Storage.java`. I think I would probably add more explanatory comments on the exception classes I've created, so other devs know what these are for.

File: src/main/java/duke/task/Deadline.java (hunk 14 above)
Comment (huizhuansam) [reply src/main/java/duke/task/Deadline.java]: Nice catch.

File: src/main/java/duke/task/Event.java (hunk 15 above)
Comment (huizhuansam) [reply src/main/java/duke/task/Event.java]: 👍🏼

**Comment by huizhuansam** [comment]
@Neha-5678 Thanks for the review!
